pkg/pillar/cmd/downloader/downloader.go

```diff
@@ -452,18 +493,66 @@ func doDownload(ctx *downloaderContext, config types.DownloaderConfig, status *t
 		return
 	}
 
-	dst, err := utils.LookupDatastoreConfig(ctx.subDatastoreConfig, config.DatastoreID)
-	if dst == nil {
-		errStr := fmt.Sprintf("Will retry when datastore available: %s",
-			err.Error())
+	// Prepend main datastore to the fallback datastore list in order
+	// to start downloading from the main datastore first
+	dsids := append([]uuid.UUID{config.DatastoreID},
+		config.FallbackDatastoreIDsList...)
+	dslist, err := prepareDatastoresList(ctx, config, dsids)
+	if err != nil {
+		errStr := fmt.Sprintf("Retry download in %v: %s failed: %s",
+			retryTime, config.Name, err)
```

Comment:
As I can see `resolveTagsToHash` function and resolve process in general also requires modification to support multiple datastores for containers. But I am not sure that it is possible to obtain sha of manifest from one datastore (i.e. docker.io), but use it to download an object from another datastore (i.e. ghcr.io). This seems to lead to conflict. WDYT, @deitch?

Reply:
I forgot to mention in the cover letter, that an OCI registry for fallback datastores is not supported. I assume that the whole async resolve mechanism works only for containers (OCI registry), so for now I would like to avoid this complexity.